util.go

```diff
@@ -11,3 +16,47 @@ var KValue = 20
 // Deprecated: This value is still the default, but changing this parameter no longer does anything.
 // Alpha is the concurrency factor for asynchronous requests.
 var AlphaValue = 3
+
+// RoutabilityBasedModeSwitching uses LocalRoutabilty events sent over the DHT's EventBus to trigger switching between
+// client and server modes.
+func RoutabilityBasedModeSwitching(dht *IpfsDHT) {
+	const modeBufSz = 256
+
+	evts := []interface{}{
+		&event.EvtLocalRoutabilityPublic{},
+		&event.EvtLocalRoutabilityPrivate{},
+		&event.EvtLocalRoutabilityUnknown{},
+	}
+
+	sub, err := dht.Host().EventBus().Subscribe(evts, eventbus.BufSize(modeBufSz))
```

Comment:
Have created a PR for the stateful event-bus at https://github.com/libp2p/go-libp2p-autonat/pull/51.

Reply:
> Not a big deal here since the events are not mutually exclusive.

You meant are mutually exclusive, right? Also, my feeling is if we're blocking with 256 routability events we probably have other issues to worry about (e.g. making autonat require more evidence before sending out events).

Reply:
are mutually exclusive* :)